Apsitvarkom.ModelActions/Validation/CleaningEventFinalizeRequestValidator.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Apsitvarkom.DataAccess;
+using Apsitvarkom.Models;
+using Apsitvarkom.Models.Public;
+using FluentValidation;
+
+namespace Apsitvarkom.ModelActions.Validation;
+
+public class CleaningEventFinalizeRequestValidator : AbstractValidator<CleaningEventFinalizeRequest>
+{
+    private readonly IPollutedLocationRepository _pollutedLocationRepository;
+
+    public CleaningEventFinalizeRequestValidator(IPollutedLocationRepository pollutedLocationRepository)
+    {
+        _pollutedLocationRepository = pollutedLocationRepository;
+
+        RuleFor(l => l.NewProgress).NotNull().InclusiveBetween(0, 100);
+        RuleFor(l => l.Id).NotNull().DependentRules(() =>
+        {
+            RuleFor(l => l.Id).MustAsync(async (req, _, context, _) =>
+            {
+                PollutedLocation? pollutedLocation;
+                try
+                {
+                    pollutedLocation = await _pollutedLocationRepository.GetByPropertyAsync(x => x.Events.Any(e => e.Id == req.Id));
+                }
+                catch (Exception)
+                {
+                    return true;
+                }
+
+                if (pollutedLocation is null)
+                    return true;
+
+                var cleaningEvent = pollutedLocation.Events.Single(x => x.Id == req.Id);
+
+                if (cleaningEvent.IsFinalized)
+                {
+                    context.AddFailure("The cleaning event is already finalized.");
+                    return true;
+                }
+
+                if (cleaningEvent.StartTime > DateTime.UtcNow)
+                    context.AddFailure("The requested cleaning event hasn't yet happened.");
+
+                if (req.NewProgress is not null && pollutedLocation.Progress > req.NewProgress)
+                    context.AddFailure($"New progress value cannot be lower than the current ({pollutedLocation.Progress}).");
+
+                return true;
+            });
```

Comment:
If you send `NewProgress = -1`, do you get both error messages? If so, should this be the expected behavior?

Reply:
good catch. this is not expected, and I've just accidentally missed it. I've previously had the progress value to be constrained just to not be higher than 100 but figured out we could better have the full 0-100 shown in swagger and didn't readjust this error message. thanks.